include/ensmallen_bits/cmaes/ipop_cmaes.hpp

```diff
@@ -0,0 +1,159 @@
+/**
+ * @file ipop_cmaes.hpp
+ * @author Marcus Edel
+ * @author Suvarsha Chennareddy
+ * @author Benjami Parellada
+ * 
+ * Definition of the IPOP Covariance Matrix Adaptation Evolution Strategy 
+ * as proposed by A. Auger and N. Hansen in "A Restart CMA Evolution 
+ * Strategy With Increasing Population Size".
+ *
+ * ensmallen is free software; you may redistribute it and/or modify it under
+ * the terms of the 3-clause BSD license.  You should have received a copy of
+ * the 3-clause BSD license along with ensmallen.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef ENSMALLEN_CMAES_IPOP_CMAES_HPP
+#define ENSMALLEN_CMAES_IPOP_CMAES_HPP
+
+#include "cmaes.hpp"
+
+namespace ens {
+
+/**
+ * IPOP CMA-ES is a variant of the stochastic search algorithm
+ * CMA-ES - Covariance Matrix Adaptation Evolution Strategy.
+ * IPOP CMA-ES, also known as CMAES with increasing population size, 
+ * incorporates a restart strategy that involves gradually increasing
+ * the population size. This approach is specifically designed to 
+ * enhance the performance of CMA-ES on multi-modal functions.
+ *
+ * For more information, please refer to:
+ *
+ * @code
+ * @INPROCEEDINGS{1554902,
+ *   author={Auger, A. and Hansen, N.},
+ *   booktitle={2005 IEEE Congress on Evolutionary Computation},
+ *   title={A restart CMA evolution strategy with increasing population size},
+ *   year={2005},
+ *   volume={2},
+ *   number={},
+ *   pages={1769-1776 Vol. 2},
+ *   doi={10.1109/CEC.2005.1554902}}
+ * @endcode
+ * 
+ * IPOP CMA-ES can optimize separable functions.  For more details, see the
+ * documentation on function types included with this distribution or on the
+ * ensmallen website.
+ *
+ * @tparam CMAESType The CMA-ES type used for the optimization. Currently, 
+ *       either CMAES or ActiveCMAES can be used.
+ */
+template<typename CMAESType = CMAES<>>
+class IPOPCMAES
+{
+ public:
+  /**
+   * Construct the IPOP CMA-ES optimizer with the given CMAES object,
+   * population factor, and maximum number of restarts. The defaults 
+   * here are not necessarily good for the given problem, so it is
+   * suggested that the values used be tailored to the task at hand.  The
+   * maximum number of iterations refers to the maximum number of points that
+   * are processed (i.e., one iteration equals one point; one iteration does not
+   * equal one pass over the dataset).
+   *
+   * @param CMAES The CMAES object.
+   * @param populationFactor The factor by which population increases
+   *      after each restart.
+   * @param maxRestarts Maximum number of restarts.
+   */
+  IPOPCMAES(const CMAESType& CMAES = CMAESType(),
```

Comment:
Both IPOP and BIPOP are restart mechanisms for the underlying CMAES. They are originally defined as extensions for the original "vanilla" CMAES, however, the restarts could be done for any variant. My intent was to have a CMAES object so the user could do the restart mechanism on whatever CMAES they wanted ("vanilla" CMAES or `ActiveCMAES`), which is why the user would input a CMAES object.

Nevertheless, I could just initialize whichever CMAES the user wants inside with the constructor you suggest. Or just ignore the different CMAES variants and just use the "vanilla" CMAES always.

What do you suggest?

Reply:
The way you described it there, does it make more sense to simply modify CMAES to add another template parameter that represents the restart mechanism?  That could simplify the code significantly.

From a user perspective, regardless of what choice you make internally, I do think it's important to make the constructor "feel" the same as the other CMAES variants, whether that is done by having IPOP-CMAES and BIPOP-CMAES be typedefs of CMAES with certain template parameters, or by holding a CMAES object internally.  So, e.g., a constructor taking each of the same arguments as the CMAES constructor (or however many of those arguments are applicable).

Hopefully that makes sense, let me know if I can clarify anything I wrote. :+1:

Reply:
> The way you described it there, does it make more sense to simply modify CMAES to add another template parameter that represents the restart mechanism? That could simplify the code significantly.

In this case, you mean adding something like `template  typename SelectionPolicyType = FullSelection,  typename TransformationPolicyType = EmptyTransformation<>,  typename RestartPolicyType = NoRestart>` instead of having what I did right now? 

Then I would need to modify the CMAES_impl.hpp `Optimize` to have something along the lines of `restartPolicy.shouldRestart()` and `restartPolicy.restart()`? Where the restart policy could either be BIPOP or IPOP or new policies.

If that's the case, I see what you mean, and it might indeed be cleaner. 

But currently, I'm using the `Optimize` from the CMAES, and when it terminates, that elicits a restart. I would have to change quite a bit of code so that instead of returning when there is a stop condition, it loops until the restarts are done.

I'm fine doing it either way, but I might need to think about how to do it.

Reply:
This could be handled by the callback feature, that said, I don't want to make the code too complex. Instead should we just create an independent IPOP implementation, that does not reuse CMAES? BIPOP could be a flag to the IPOP implementation. What do you think?

Reply:
Right, I had sort of been thinking about adding a `RestartPolicyType` template parameter to CMAES.  But an independent IPOP implementation would maybe be the better thing.  Honestly I don't have a huge opinion either way (and I don't deeply understand the algorithm), my perspective is entirely user-facing.  So as long as the class "feels the same" as other optimizers to use, I'm happy. :smile:

Reply:
> BIPOP could be a flag to the IPOP implementation. 

I can definitely reduce both into one, with a flag to choose. 

As for the code, I'm also fine doing either way, I feel like creating an instance of CMAES inside the IPOP and then use `Optimize` might be the best though. This way the user can control all the parameters in the same fashion, while also not duplicating code, and any improvement we do to the CMAES will be reflected in the IPOP (and BIPOP).

Reply:
@zoq @rcurtin I refactored it into one code only (POP-CMA-ES) where with a flag it launches BIPOP or IPOP. I also tried to make it as similar to the other CMAES optimizers. However, now instead of passing it an CMAES object, I just initializes the CMAES class inside and use it.

Let me know if you prefer this version, and if you do, i'll write out the documentation for this and should be pretty much ready.

Reply:
I like it, thanks for taking the effort to do that. Two quick notes:

 * Feel free to use underscores in the class names, like `POP_CMAES`; this can make the class name a little more decipherable. :)
 * If you want to keep the BIPOP and IPOP versions as separate optimizers, you could make the `useBIPOP` parameter into a template parameter and use a `using` statement to define the `IPOPCMAES` and `BIPOPCMAES` classes. Up to you whether you want to do that---it really hinges on whether you want to provide to the user a single class `POPCMAES` or two classes. (I guess two classes would be more in line with how ensmallen is structured, with one optimizer per class, but either way is fine with me.)

Thanks again for the effort; at least from my side I have no other major comments. :+1:

Reply:
All the comments make sense to me.

Reply:
Done the suggestions @rcurtin!

Going to mark this no longer as draft